### PR TITLE
Warning if categorical crossentropy is used for 2 labels

### PR DESCRIPTION
**Pull request by old-school-kid**
solves #17029 
[Solution colab notebook](https://colab.research.google.com/drive/1nUBPGfLs2CGs6LjA2_luscuWadZ0kwnQ?usp=sharing)
Raised a warning and not an error. 

Also `CategoricalHinge` loss could have the same problem but apparently the documentation itself cites an 1D tensor as an example. It should be noted that the example given in the doc returns `1.4` as answer while the correct answer is `1.3`

cc @haifeng-jin

### CODE REVIEW

**Review by haifeng-jin — CHANGES_REQUESTED** [review]
Thanks for the PR! Would you please add a unit test for the warning?

**Comment by old-school-kid** [comment]
@haifeng-jin 
Sorry for the delay. Please take a look now.

File: keras/losses_test.py [hunk 1]
```diff
@@ -1794,6 +1796,16 @@ def test_ragged_tensors_ragged_sample_weights(self):
...
         self.assertAlmostEqual(self.evaluate(loss), 0.3181, 3)
 
+    def test_binary_labels(self):
+        # raise a warning if the shape of y_true and y_pred are all (None, 1).
+        # categorical_crossentropy shouldn't be used with binary labels.
+        with warnings.catch_warnings(record=True) as w:
+            warnings.simplefilter("always")
+            cce_obj = losses.CategoricalCrossentropy()
+            cce_obj(tf.constant([[1.0], [0.0]]), tf.constant([[1.0], [1.0]]))
+            assert issubclass(w[-1].category, SyntaxWarning)
```
Comment (haifeng-jin) [inline keras/losses_test.py:1806]: Let's use `self.assertIsInstance(w[-1].category, SyntaxWarning)`.

File: keras/losses_test.py [hunk 2]
```diff
@@ -1794,6 +1796,16 @@ def test_ragged_tensors_ragged_sample_weights(self):
...
+            assert "Recieved an one-dimensional output..*" in str(w[-1].message)
```
Comment (haifeng-jin) [inline keras/losses_test.py:1807]: Let's use `self.assertIn(your_string, w[-1].message)`.

File: keras/losses.py [hunk 3]
```diff
@@ -1959,6 +1960,15 @@ def categorical_crossentropy(
     y_true = tf.cast(y_true, y_pred.dtype)
     label_smoothing = tf.convert_to_tensor(label_smoothing, dtype=y_pred.dtype)
 
+    if y_pred.shape[-1] == 1:
+        warnings.warn(
+            "Recieved an one-dimensional output. "
+            "Consider using binary crossentropy "
+            "instead of categorical crossentropy "
+            "if you have only 2 labels",
```
Comment (haifeng-jin) [inline keras/losses.py:1968]: Usually we do the strings in the following format:
```py
f"Expected something to be something. Received: {something}."
```
So my suggestion would be:
```py
"Expected the tensor's shape passed to 'categorical_crossentropy' to be (batch_size, n_classes), "
f"where n_classes > 1. Received: y_pred.shape={y_pred.shape}. "
"Consider using 'binary_crossentropy' if you only have 2 classes."
```

See if @fchollet have comments on this part.

**Review by haifeng-jin — CHANGES_REQUESTED** [review]
Thanks for the update!

**Review by haifeng-jin — APPROVED** [review]
Looks good to me! We will see @fchollet review before merging.

File: keras/losses.py [hunk 4]
```diff
@@ -1959,6 +1960,15 @@ def categorical_crossentropy(
     y_true = tf.cast(y_true, y_pred.dtype)
     label_smoothing = tf.convert_to_tensor(label_smoothing, dtype=y_pred.dtype)
 
+    if y_pred.shape[-1] == 1:
+        warnings.warn(
+            "Expected the tensor's shape passed to 'categorical_crossentropy' "
+            "to be (batch_size, n_classes), "
+            f"where n_classes > 1. Received: y_pred.shape={y_pred.shape}. "
+            "Consider using 'binary_crossentropy' if you only have 2 classes.",
+            SyntaxWarning,
```
Comment (fchollet) [inline keras/losses.py:1969]: Add `stacklevel=2` argument

File: keras/losses.py [hunk 5]
```diff
@@ -1959,6 +1960,15 @@ def categorical_crossentropy(
     y_true = tf.cast(y_true, y_pred.dtype)
     label_smoothing = tf.convert_to_tensor(label_smoothing, dtype=y_pred.dtype)
 
+    if y_pred.shape[-1] == 1:
+        warnings.warn(
+            "Expected the tensor's shape passed to 'categorical_crossentropy' "
+            "to be (batch_size, n_classes), "
+            f"where n_classes > 1. Received: y_pred.shape={y_pred.shape}. "
+            "Consider using 'binary_crossentropy' if you only have 2 classes.",
```
Comment (fchollet) [inline keras/losses.py:1968]:
```
"In loss categorical_crossentropy, expected "
"y_pred.shape to be (batch_size, num_classes) "
f"with num_classes > 1. Received: y_pred.shape={y_pred.shape}. "
"Consider using 'binary_crossentropy' if you only have 2 classes.",
```

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

**Comment by haifeng-jin** [comment]
@old-school-kid 
I have created https://github.com/old-school-kid/keras/pull/3 .
You can merge it and the tests are fixed.

**Comment by haifeng-jin** [comment]
Hi @old-school-kid , the PR has been merged. Thank you again for your contribution!
Would you help us answer the following questions so that we can improve the contributor experience?
* What editor did you use to edit the code?
* Did you find it hard to setup the environment, run the tests, or format the code?

Any other feedback are also appreciated.
